### PR TITLE
DOC: Try add a section on NEP 50 to migration guide

### PR DESCRIPTION
**Pull request by seberg**
It is tricky, because there are many corner cases, but I don't want to list them.  More guidance on what to do could be nice, but I am not sure what to say beyond "convert to Python scalar".

### CODE REVIEW

**Comment by mhvk** [comment]
I think mentioning the corner cases is a benefit for ourselves at least! And are there that many? The ones that I am aware of are the comparisons and now true_divide. The section here links to the example table in NEP 50 - we could add things below that as "Implementation details" - that this is done in the ufunc machinery and that exceptions are made for some logical cases. Perhaps better would be an addition to the user documentation about promotion, but I worry that we'll let perfect be the enemy of good enough...

**Comment by MatthewFlamm** [comment]
Thanks for adding in all this further clarification/discussion to the NEP and the migration guide. The description makes sense to me at a high level and would have alerted me that this change would cause something like #25639.  A few suggestions/observations.

I don't know what is aimed in the migration vs. release notes, but a few words and a link to the NEP or the migration guide in the release notes would be helpful to know that this behavior changed.

Do you include versions of implementation in the NEPs?  I see that the version where ability to control the behavior is listed in NEP 50.  This part in particular:

> As of the NumPy 1.24.x series, NumPy has preliminary and limited support to test this proposal.

So, users may think that this is the current state in numpy 2.0.  I would suggest clarifying what happens in v2.0 in the NEP itself.

**Review by rgommers — APPROVED** [review]
Thanks @seberg! This looks well-written and with the right level of detail for this guide. Good to go I think.

File: doc/source/numpy_2_0_migration_guide.rst [hunk 1]
```diff
@@ -8,6 +8,41 @@ This document contains a set of instructions on how to update your code to
...
+The largest backwards compatibility change of this is that it means that
+the precision of scalars is now preserved consistently.
+Two examples are:
+* ``np.float32(3) + 3.`` now returns a float32 when it previously returned
+  a float64.
+* ``np.array([3], dtype=np.float32) + np.float64(3)`` will now return a float64
+  array.  (The higher precision of the scalar is not ignored.)
+
+For floating point values, this can lead to lower precision results when
+working with scalars.  For integers errors or overflows are possible.
```
Comment (mattip) [inline doc/source/numpy_2_0_migration_guide.rst:32]:
```suggestion
working with scalars.  For integers, errors or overflows are possible.
```

File: doc/source/numpy_2_0_migration_guide.rst [hunk 2]
```diff
@@ -8,6 +8,41 @@ This document contains a set of instructions on how to update your code to
 work with Numpy 2.0.
 
 
+.. _migration_promotion_changes:
+
+Changes to NumPy data type promotion
+=====================================
+
+One of the larger changes in the NumPy 2.0 were the changes to promotion
+as per `NEP 50 <NEP50_>`.
```
Comment (mattip) [inline doc/source/numpy_2_0_migration_guide.rst:17]:
```suggestion
NumPy 2.0 changes promotion (the result of combining dissimilar data types)
as per `NEP 50 <NEP50_>`.
```

File: doc/source/numpy_2_0_migration_guide.rst [hunk 3]
```diff
@@ -8,6 +8,41 @@ This document contains a set of instructions on how to update your code to
...
+To solve this, you may cast explicitly.  Very often, it may also be a good
+solution to ensure you are working with Python scalars via ``int()``,
+``float()``, or ``numpy_scalar.item()``.
+
+To track down changes, you can try to use for testing purposes only
+(use ``warnings.simplefilter`` to raise it as an error for a traceback)::
+
+  np._set_promotion_state("weak_and_warn")
+
+which attempts to give warnings about changed behavior.  Unfortunately,
```
Comment (mattip) [inline doc/source/numpy_2_0_migration_guide.rst:43]:
```suggestion
To track down changes, you can enable emitting warnings for changed behavior
(use ``warnings.simplefilter`` to raise it as an error for a traceback)::

  np._set_promotion_state("weak_and_warn")

which is useful during testing. Unfortunately,
```

**Review by mattip — COMMENTED** [review]
A few small nits.

**Comment by seberg** [comment]
Thanks, looked good, adopted the suggestions.  As to the other point:  Yeah, we never had any decent docs on promotion and we should have them somewhere and note the exceptions there. (they are now already mentioned in the NEP)

**Comment by mattip** [comment]
Thanks @seberg

**Pull request merged**